### DISCUSSION TITLE
Integrating existing c++/arduino library into an fprime component. Iridium rockBLOCK 9603 satellite module.

### DISCUSSION THREAD

**Discussion by markospacetech** [nasa/fprime#1019] [Q&A] [answered]
Hello all,

I am currently attempting to create an f prime component to interface with the rockBLOCK 9603 Iridium sat-com module. 

I have been using the module successfully with this Arduino/c++  [library](https://github.com/mikalhart/IridiumSBD )

My plan is to analyze the f prime GPS example component and follow the same pattern adding functions to send commands to the module, utilizing the functions from the Arduino library.  Still not sure how to setup the topology / serial IO however. 

I come here to see if anyone has tried something similar or can give some insight into how to better approach this task. F prime seems like an amazing project/framework and I'm looking forward to learning and leveraging it as much as possible!

Cheers

*Platform is Linux with Ubuntu 20 for ground station/uplink and RPI OS(3 B+) for downlink.

**Comment by timcanham** [comment, marked as answer]
If you can compile against the Arduino libraries on your platform, the way we've done it in the past it to wrap the API in a component where the port handlers call the API. Is this meant to act as a radio uplink/downlink to a ground station somewhere?

**Reply by markospacetech** [reply]
Ok I think I understand. So rather than importing f prime serial drivers and passing a [buffer ](https://github.com/fprime-community/gps-tutorial/blob/cf26f552c8fdfe1bf0c0fd20b7b3cd327b62f396/GpsApp/Gps/GpsComponentImpl.cpp#L73) , I can just wrap a handler for receiving and a handler for sending commands around the module's C++ API?

The module will be used for ground station uplink communicating with another module acting as radio downlink.

**Comment by SterlingPeet** [comment]
Just as an FYI, F Prime is very challenging to fit into an ATmega Arduino, even the 2560.  I have done it, but it does require the use of the XMEM external SRAM hardware extension.  Even then it can only support big parts like ATmega128 or ATmega2560, and only for very small deployments.

If you are using an ARM based Arduino like a Teensy, it should be a bit easier to fit into the platform.

**Reply by markospacetech** [reply]
Noted thank you, I have also used the module with a RPI 3 B+ connected to the GPIO pins. I will be sticking with that setup moving forward.

**Comment by timcanham** [comment]
If you want to use it as a ground interface, you can write a component that will connect to the ground ports of the [GroundInterface](https://github.com/nasa/fprime/tree/devel/Svc/GroundInterface) component.

Your component would replace the [SocketIpDriver](https://github.com/nasa/fprime/tree/devel/Drv/SocketIpDriver) used in the Ref demonstration.

The additional step beyond that would be to write an adapter for the ground system to take packets off the ground radio and convert them back to the format the ground system ingests.

Tagging @LeStarch to point to a tutorial on how to write the ground system adapter.

**Reply by LeStarch** [reply]
The ground interface component has been broken in to "Framer" and "Deframer" pieces.  If you want to interface with it, you need to write a component that has the port interface of Drv/ByteStreamDriverModel.  This provides a stream of bytes that will be deframed, and accepts a stream of bytes to be transmitted.

TcpClient, TcpServer, and Udp are examples.

**Reply by markospacetech** [reply]
So the library I was hoping to use is highly coupled with arduino libraries. There is also a python library that works well with the raspberry pi however. Would it be reasonable to use  a python library or do I need to make a c++ driver or an extension of the f prime serial driver?

**Reply by timcanham** [reply]
The typical pattern is to let the serial driver only talk to the serial port, not get involved with knowing what devices are on the other side of the the UART port. There is no requirement to use a serial port component at all - if you have a C/C++ library that does its own I/O on a UART port, just wrap it in a component. Since F Prime is a C++ framework, you could hypothetically use python by making system calls to execute the python, but that sounds complicated. A C++ sounds better.

**Reply by Joshua-Anderson** [reply]
One other suggestion is to wrap the python library in some sort of external python program with a socket interface. You could then write a F' component that reads/writes from the socket, which would talk to your python program, which would use the python library to write to the radio. This would have a lower performance penalty than running the python library within the C++component and would probably be less work than re-writing the library in C++

**Reply by markospacetech** [reply]
@Joshua-Anderson  I like that solution. For the F' component, could the TcpClient be used by specifying the py server port and then just sending/receiving a byte stream?

**Comment by Joshua-Anderson** [comment]
Yep, TCP or UDP driver should do the trick. Named pipes might also be a good option.

**Reply by Joshua-Anderson** [reply]
You set it when you call `configure` on the client. Take a look at the [component documentation](https://github.com/nasa/fprime/blob/devel/Drv/TcpClient/docs/sdd.md) for an example!
